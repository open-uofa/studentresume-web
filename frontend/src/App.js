import React from 'react';
import { BrowserRouter, Routes, Route, useNavigate } from "react-router-dom";
import FormPage from './components/FormPage';
import Mainpage from './components/Mainpage';
import ThemePage from './components/ThemePage';
import PreviewPage from './components/PreviewPage';
import { validateForm } from './actions/forms';
import ReviewPage from './components/ReviewPage';
import FormPageForMobile from './components/FormPageForMobile';
import AboutPage from './components/AboutPage';

const App = () => {
  // we store user data in state as global variable
  // const state = require("./constants/defaultInfo.json");
  const [state, setState] = React.useState(require("./constants/default.json"));
  // const [themeState, setThemeState] = React.useState(require("./constants/defaultTheme.json"));
  const [sectionComplete, setSectionComplete] = React.useState(require("./constants/completion.json"));

  const [isFormValid, setIsFormValid] = React.useState(false);

  const [uploading, setUploading] = React.useState(false);

  const onUpdateChange = (sectionTitle, index = 0, fieldName, value) => {

    if(fieldName === 'highlights') {
      // convert value to array
      value = value.split('\n');
    }
    console.log(value)
    if (sectionTitle === 'basics') {
      // update state
      setState((prevState) => ({
        ...prevState,
        [sectionTitle]: {
          ...prevState[sectionTitle],
          [fieldName]: value
        }
      }));
    }
    else {
      // update state
      setState((prevState) => ({
        ...prevState,
        [sectionTitle]: [
          ...prevState[sectionTitle].slice(0, index),
          {
            ...prevState[sectionTitle][index],
            [fieldName]: value
          },
          ...prevState[sectionTitle].slice(index + 1)

        ]
      }));
    }
    console.log(state)
  }

  const onAddEntry = (sectionTitle) => {
    // update state
    setState((prevState) => ({
      ...prevState,
      [sectionTitle]: [
        ...prevState[sectionTitle],
        require("./constants/default.json")[sectionTitle][0]
      ]
    }));
  }

  const onDeleteEntry = (sectionTitle, index) => {
    // update state
    setState((prevState) => ({
      ...prevState,

      [sectionTitle]: (prevState[sectionTitle].length == 1) ?
        [
          require("./constants/default.json")[sectionTitle][0]
        ] : [
          ...prevState[sectionTitle].slice(0, index),
          ...prevState[sectionTitle].slice(index + 1)
        ]
    }));
  }

  const onAddKeyword = (sectionTitle, index, key, keyword) => {
    // update state
    setState((prevState) => ({
      ...prevState,
      [sectionTitle]: [
        ...prevState[sectionTitle].slice(0, index),
        {
          ...prevState[sectionTitle][index],
          [key]: [
            ...prevState[sectionTitle][index][key],
            keyword
          ]
        },
        ...prevState[sectionTitle].slice(index + 1)
      ]
    }));
  }

  const onDeleteKeyword = (sectionTitle, index, key, keywordIndex) => {
    // update state
    setState((prevState) => ({
      ...prevState,
      [sectionTitle]: [
        ...prevState[sectionTitle].slice(0, index),
        {
          ...prevState[sectionTitle][index],
          keywords: [
            ...prevState[sectionTitle][index].keywords.slice(0, keywordIndex),
            ...prevState[sectionTitle][index].keywords.slice(keywordIndex + 1)
          ]
        },
        ...prevState[sectionTitle].slice(index + 1)
      ]
    }));
  }

  const onUpdateSectionComplete = (sectionTitle, value) => {
    setSectionComplete((prevState) => ({
      ...prevState,
      [sectionTitle]: value
    }));
  }

  const onLoadingLocalStorage = () => {
    // load from local storage
    const localState = JSON.parse(localStorage.getItem("state"));
    const localSectionComplete = JSON.parse(localStorage.getItem("sectionComplete"));
    if (localState) {
      setState(localState);
    }
    if (localSectionComplete) {
      setSectionComplete(localSectionComplete);
    }
  }

  const onSavingLocalStorage = () => {
    // save to local storage
    localStorage.setItem("state", JSON.stringify(state));
    localStorage.setItem("sectionComplete", JSON.stringify(sectionComplete));
  }

  const cleanLocalStorage = () => {
    localStorage.removeItem("state");
    localStorage.removeItem("sectionComplete");
  }

  React.useEffect(() => {
    onLoadingLocalStorage();
  }, []);

  return (
    <div>
      {uploading ? (
        <div className="App">
          {/* TODO: design uploading */}
          <h1>Uploading...</h1>
        </div>
      ) : (
        <BrowserRouter>
          <Routes>
            <Route path="/" element={<Mainpage setUploading={setUploading} setState={setState} onSavingLocalStorage={onSavingLocalStorage} />} />
            <Route path="/form" element={<FormPage
              userData={state}
              sectionComplete={sectionComplete}
              onUpdateChange={onUpdateChange}
              onAddEntry={onAddEntry}
              onDeleteEntry={onDeleteEntry}
              onAddKeyword={onAddKeyword}
              onDeleteKeyword={onDeleteKeyword}
              setUploading={setUploading}
              onUpdateSectionComplete={onUpdateSectionComplete}
              onLoadingLocalStorage={onLoadingLocalStorage}
              onSavingLocalStorage={onSavingLocalStorage}
              cleanLocalStorage={cleanLocalStorage} />}
            />
            <Route path="/form-mobile" element={<FormPageForMobile
              userData={state}
              sectionComplete={sectionComplete}
              onUpdateChange={onUpdateChange}
              onAddEntry={onAddEntry}
              onDeleteEntry={onDeleteEntry}
              onAddKeyword={onAddKeyword}
              onDeleteKeyword={onDeleteKeyword}
              setUploading={setUploading}
              onUpdateSectionComplete={onUpdateSectionComplete}
              onLoadingLocalStorage={onLoadingLocalStorage}
              onSavingLocalStorage={onSavingLocalStorage}
              cleanLocalStorage={cleanLocalStorage} />}
            />
            <Route path="/review" element={<ReviewPage userData={state} setUploading={setUploading} onLoadingLocalStorage={onLoadingLocalStorage} />} />
            <Route path="/theme" element={<ThemePage userData={state} setUploading={setUploading} onSavingLocalStorage={onSavingLocalStorage} />} />
            <Route path="/preview" element={<PreviewPage userData={state} />} />
            <Route path="/about" element={<AboutPage />} />
          </Routes>
        </BrowserRouter>
      )}
    </div>
  );
}

export default App;
