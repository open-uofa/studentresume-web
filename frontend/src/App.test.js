import App from './App';
import React from 'react';
import renderer from 'react-test-renderer'

test("First snapshot test", () => {
  const component = renderer.create(
    <App />
  );

  let tree = component.toJSON();

  expect(tree).toMatchSnapshot();
})

test("State Testing", () => {

  const tempState = require("./constants/example.json");

  expect(tempState[0]).toBe(App.state);

}) 