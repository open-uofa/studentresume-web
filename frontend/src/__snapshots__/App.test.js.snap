// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First snapshot test 1`] = `
<div>
  <div
    className="main_page"
  >
    <h2
      style={
        Object {
          "alignItems": "left",
          "fontWeight": "normal",
          "marginLeft": "2%",
          "marginTop": "1.5%",
        }
      }
    >
      Student
      <b>
        Resume
      </b>
    </h2>
    <p
      href="#doc_link"
      style={
        Object {
          "marginRight": "2%",
          "marginTop": "1.5%",
          "position": "absolute",
          "right": "2em",
        }
      }
    >
      Never heard of JSON? Click 
      <a
        data-testid="docJSON"
        href="https://www.json.org/json-en.html"
      >
        here
      </a>
       for documentation
    </p>
    <div
      className="MuiBox-root css-1j4a9v9"
    >
      <h1
        style={
          Object {
            "color": "#464646",
            "fontSize": "50px",
            "margin": "2%",
            "marginBottom": "4%",
            "textAlign": "center",
          }
        }
      >
        How would you like to start?
      </h1>
      <div
        className="MuiBox-root css-aon0ne"
      >
        <a
          href="/form"
          onClick={[Function]}
          style={
            Object {
              "textDecoration": "none",
            }
          }
        >
          <div
            className="MuiBox-root css-1obf42f"
            data-testid="goToForm"
          >
            <div
              className="MuiBox-root css-ggl79w"
            >
              <img
                alt=""
                height="100%"
                src="resume.png"
                width="75%"
              />
            </div>
            <div
              className="MuiBox-root css-1wsrftx"
            >
              <h3
                style={
                  Object {
                    "color": "#464646",
                    "fontSize": "25px",
                    "margin": "2%",
                    "marginTop": "7%",
                    "textAlign": "center",
                  }
                }
              >
                Create a New Resume
              </h3>
              <h4
                style={
                  Object {
                    "color": "#A9A9A9",
                    "fontSize": "16px",
                    "margin": "2%",
                    "marginTop": "4%",
                    "textAlign": "center",
                  }
                }
              >
                We'll guide you through each resume section!
              </h4>
            </div>
          </div>
        </a>
        <div
          className="MuiBox-root css-con467"
        >
          <div
            className="MuiBox-root css-bhj89f"
          >
            <div
              className="center"
            >
              <div
                className="drag_drop_text2"
                style={
                  Object {
                    "color": "#464646",
                    "fontSize": "15px",
                    "margin": "2%",
                    "marginBottom": "4%",
                    "textAlign": "center",
                  }
                }
              >
                Only JSON files are allowed
              </div>
              <div
                className="drag_drop"
                data-testid="dragdrop"
                onBlur={[Function]}
                onClick={[Function]}
                onDragEnter={[Function]}
                onDragLeave={[Function]}
                onDragOver={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                role="presentation"
                tabIndex={0}
              >
                <input
                  accept=".json,.json"
                  multiple={true}
                  onChange={[Function]}
                  onClick={[Function]}
                  style={
                    Object {
                      "display": "none",
                    }
                  }
                  tabIndex={-1}
                  type="file"
                />
                <img
                  alt=""
                  height="70%"
                  src="file_icon.png"
                  width="30%"
                />
                <h5
                  style={
                    Object {
                      "color": "#464646",
                      "fontSize": "15px",
                      "margin": "2%",
                      "marginBottom": "4%",
                      "textAlign": "center",
                    }
                  }
                >
                  Drag and Drop your files here
                </h5>
              </div>
            </div>
          </div>
          <div
            className="MuiBox-root css-1wsrftx"
          >
            <h3
              style={
                Object {
                  "color": "#464646",
                  "fontSize": "25px",
                  "margin": "2%",
                  "marginTop": "7%",
                  "textAlign": "center",
                }
              }
            >
              Upload my JSON File
            </h3>
            <h4
              style={
                Object {
                  "color": "#A9A9A9",
                  "fontSize": "16px",
                  "margin": "2%",
                  "marginTop": "4%",
                  "textAlign": "center",
                }
              }
            >
              We'll move everything to the fill out form!
            </h4>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
